paddle/math/float16.h

```diff
@@ -0,0 +1,714 @@
+    return *this;
+  }
+
+  PADDLE_HOSTDEVICE inline float16& operator=(int64_t val) {
+    float16 res = fp16_impl::float_to_half_rn(static_cast<float>(val));
+    x = res.x;
+    return *this;
+  }
+
+  PADDLE_HOSTDEVICE inline float16& operator=(uint64_t val) {
+    float16 res = fp16_impl::float_to_half_rn(static_cast<float>(val));
+    x = res.x;
+    return *this;
+  }
+
+  PADDLE_HOSTDEVICE inline float16& operator=(float val) {
+    float16 res = fp16_impl::float_to_half_rn(val);
+    x = res.x;
+    return *this;
+  }
+
+  PADDLE_HOSTDEVICE inline float16& operator=(double val) {
+    float16 res = fp16_impl::float_to_half_rn(static_cast<float>(val));
+    x = res.x;
+    return *this;
+  }
+
+#ifdef PADDLE_CUDA_FP16
+  PADDLE_HOSTDEVICE inline operator half() const {
+#if CUDA_VERSION >= 9000
+    __half_raw h;
+    h.x = x;
+    return half(h);
+#else
+    half h;
+    h.x = x;
+    return h;
+#endif  // CUDA_VERSION >= 9000
+  }
+#endif  // PADDLE_CUDA_FP16
+
+  PADDLE_HOSTDEVICE inline operator Eigen::half() const {
+    Eigen::half h;
+    h.x = x;
+    return h;
+  }
+
+#if defined(PADDLE_NEON) && defined(PADDLE_ARM_FP16) && \
+    (PADDLE_GNUC_VER >= 61 || PADDLE_CLANG_VER >= 34)
+  PADDLE_HOSTDEVICE inline operator float16_t() const {
+    float16 h = *this;
+    return *reinterpret_cast<float16_t*>(&h);
+  }
+#endif
+
+  PADDLE_HOSTDEVICE inline explicit operator bool() const {
+    return (x & 0x7fff) != 0;
+  }
+
+  PADDLE_HOSTDEVICE inline explicit operator int8_t() const {
+    return static_cast<int8_t>(fp16_impl::half_to_float(*this));
+  }
+
+  PADDLE_HOSTDEVICE inline explicit operator uint8_t() const {
+    return static_cast<uint8_t>(fp16_impl::half_to_float(*this));
+  }
+
+  PADDLE_HOSTDEVICE inline explicit operator int16_t() const {
+    return static_cast<int16_t>(fp16_impl::half_to_float(*this));
+  }
+
+  PADDLE_HOSTDEVICE inline explicit operator uint16_t() const {
+    return static_cast<uint16_t>(fp16_impl::half_to_float(*this));
+  }
+
+  PADDLE_HOSTDEVICE inline explicit operator int32_t() const {
+    return static_cast<int32_t>(fp16_impl::half_to_float(*this));
+  }
+
+  PADDLE_HOSTDEVICE inline explicit operator uint32_t() const {
+    return static_cast<uint32_t>(fp16_impl::half_to_float(*this));
+  }
+
+  PADDLE_HOSTDEVICE inline explicit operator int64_t() const {
+    return static_cast<int64_t>(fp16_impl::half_to_float(*this));
+  }
+
+  PADDLE_HOSTDEVICE inline explicit operator uint64_t() const {
+    return static_cast<uint64_t>(fp16_impl::half_to_float(*this));
+  }
+
+  PADDLE_HOSTDEVICE inline explicit operator float() const {
+    return fp16_impl::half_to_float(*this);
+  }
+
+  PADDLE_HOSTDEVICE inline explicit operator double() const {
+    return static_cast<double>(fp16_impl::half_to_float(*this));
+  }
+};
+
+// Arithmetic operators
+#if defined(PADDLE_CUDA_FP16) && defined(__CUDA_ARCH__) && __CUDA_ARCH__ >= 530
+__device__ inline float16 operator+(const float16& a, const float16& b) {
```

Comment:
Do we need to define these CUDA device operations?
We can use the half type directly in CUDA's Kernel.

Reply:
Please refer to https://github.com/PaddlePaddle/Paddle/pull/5851/files for the different support provided by different versions of CUDA to half type.

By defining these CUDA device operations here along with the implicit conversion operator between our float16 and half, we can run the following code on CUDA < 9.0 (have tested on our nvidia-docker image with CUDA 8.0):

```
namespace paddle {
__global__ void() {
  half a, b, c;
  // correct for cuda >= 7.5 if defined inside paddle namespace
  // gives compiler error if not put in paddle namespace for cuda < 9.0
  c = a + b;
}
}
```

So these device operations make our code using cuda half data type arithmetic operations easy to write and compatible with all CUDA >= 7.5.

Reply:
> However, if we call c = a + b with a, b, c all being half data type, it is much less efficient compared to c = __hadd(a, b) because of all the unnecessary conversions performed.

Yeah, implicit conversion is dangerous, the declaration of the conversion function needs to add `explicit`.

Reply:
So, for now, there is two way to implementations for the FP16 Kernel in CUDA.
1. Use `half` and operator like `c = __hadd(a, b)` when CUDA < 9.0, this kernel is also work when CUDA >= 9.0
2. Use `half` and operator like `c = a + b` when CUDA >= 9.0;
For the first, if I am an outside contributor, I am not familiar with the type definition of paddle.
For the second, if we need those kernels write for CUDA >= 9.0 work well when CUDA < 9.0, I think the opinion in your comment is better(define `operator+` for cuda half when CUDA < 9.0).